Update to Freyja 1.3.12

jl32587 opened the pull request:
<!--
Thank your for contributing to the Staph-B community!
Please fill in the appropriate checklist below and delete whatever is not relevant.

These recommendations are not meant as a means to keep people from contributing. 
If you need assistance with the following measures, submit your pull request anyway, and we will work with you...
Or you can contact us beforehand via slack or by sumitting a pull request https://github.com/StaPH-B/docker-builds/issues

Documentation on how to contribute can be found at https://staphb.org/docker-builds/contribute/
Documentation for how to create a Dockerfile can be found at https://staph-b.github.io/docker-builds/make_containers
A recommended Dockerfile template can be found at https://github.com/StaPH-B/docker-builds/blob/master/dockerfile-template/Dockerfile


Please replace all '[ ]' with '[X]' to demonstrate completion.
-->

Pull Request (PR) checklist:
- [X] Include a description of what is in this pull request in this message.
- [X] The dockerfile successfully builds to a test target for the user creating the PR. (i.e. `docker build --tag samtools:1.15test --target test docker-builds/samtools/1.15` )
- [X] Directory structure as name of the tool in lower case with special characters removed with a subdirectory of the version number (i.e. `spades/3.12.0/Dockerfile`)
   - [X] (optional) All test files are located in same directory as the Dockerfile (i.e. `shigatyper/2.0.1/test.sh`)
- [X] Create a simple container-specific [README.md](https://github.com/StaPH-B/docker-builds/blob/master/.github/workflow-templates/readme-template.md) in the same directory as the Dockerfile (i.e. `spades/3.12.0/README.md`)
   - [X] If this README is longer than 30 lines, there is an explanation as to why more detail was needed
- [X] Dockerfile includes the recommended [LABELS](https://github.com/StaPH-B/docker-builds/blob/master/dockerfile-template/Dockerfile#L8-L18) 
- [X] Main [README.md](https://github.com/StaPH-B/docker-builds/blob/master/README.md) has been updated to include the tool and/or version of the dockerfile(s) in this PR
- [X] [Program_Licenses.md](https://github.com/StaPH-B/docker-builds/blob/master/Program_Licenses.md) contains the tool(s) used in this PR and has been updated for any missing


<!-- If this PR is for something else, please add extra descriptions -->

There's a new version of Freyja out. According to https://github.com/andersen-lab/Freyja/releases/tag/v1.3.12,

Uses updates downloaded from repo by default (updated and recorded with github action). Adds weekly plot interval, reverses legend order to match plot. Also adds flexibility to add lineages that aren't observed in public data (i.e. gisaid-only).

erinyoung commented on freyja/1.3.12/Dockerfile:
```diff
@@ -0,0 +1,63 @@
+LABEL description="Freyja is a tool to recover relative lineage abundances from mixed SARS-CoV-2 samples from a sequencing dataset (BAM aligned to the Hu-1 reference)"
+LABEL website="https://github.com/andersen-lab/Freyja"
+LABEL license="https://github.com/andersen-lab/Freyja/blob/main/LICENSE"
+LABEL maintainer1="Kevin Libuit"
+LABEL maintainer1.email="kevin.libuit@theiagen.com"
```
This looks great! 

This dockerfile is going to fail testing, though, because we need to adjust maintainer1 to maintainer and maintainer1.email to maintainer.email (the rest can probably stay the same).

erinyoung commented on freyja/1.3.12/Dockerfile:
```diff
@@ -0,0 +1,63 @@
+
+# Grab test data
+RUN wget -O /data/Freyja_WWSC2.bam https://github.com/StaPH-B/docker-builds/blob/master/freyja/1.3.4/tests/Freyja_WWSC2.bam?raw=true -O /data/Freyja_WWSC2.bam && \
+ wget -P /data https://raw.githubusercontent.com/StaPH-B/docker-builds/master/freyja/1.3.4/tests/Freyja_depths.tsv && \
+ wget -P /data https://raw.githubusercontent.com/StaPH-B/docker-builds/master/freyja/1.3.4/tests/Freyja_variants.tsv && \
+ wget -P /data https://raw.githubusercontent.com/StaPH-B/docker-builds/master/freyja/1.3.4/tests/nCoV-2019.reference.fasta
+
```
It looks like the files you're using for comparison were built under a different Freyja version.

erinyoung commented:
This was the part of the test that failed:

```
#11 [test 3/4] RUN cmp /data/test_variants.tsv /data/Freyja_variants.tsv && cmp /data/test_depths.tsv /data/Freyja_depths.tsv
#0 0.065 /data/test_variants.tsv /data/Freyja_variants.tsv differ: byte 139, line 1
#11 ERROR: process "/usr/local/bin/_dockerfile_shell.sh cmp /data/test_variants.tsv /data/Freyja_variants.tsv && cmp /data/test_depths.tsv /data/Freyja_depths.tsv" did not complete successfully: exit code: 1
------
 > [test 3/4] RUN cmp /data/test_variants.tsv /data/Freyja_variants.tsv && cmp /data/test_depths.tsv /data/Freyja_depths.tsv:
#0 0.065 /data/test_variants.tsv /data/Freyja_variants.tsv differ: byte 139, line 1
------
WARNING: local cache import at /tmp/.buildx-cache-freyja not found due to err: could not read /tmp/.buildx-cache-freyja/index.json: open /tmp/.buildx-cache-freyja/index.json: no such file or directory
Dockerfile:62
--------------------
  60 |     
  61 |     # Check validity of outputs
  62 | >>> RUN cmp /data/test_variants.tsv /data/Freyja_variants.tsv && cmp /data/test_depths.tsv /data/Freyja_depths.tsv
  63 |     RUN grep "Omicron" /data/test_demix.tsv && echo done
  64 |     
--------------------
ERROR: failed to solve: process "/usr/local/bin/_dockerfile_shell.sh cmp /data/test_variants.tsv /data/Freyja_variants.tsv && cmp /data/test_depths.tsv /data/Freyja_depths.tsv" did not complete successfully: exit code: 1
Error: buildx failed with: ERROR: failed to solve: process "/usr/local/bin/_dockerfile_shell.sh cmp /data/test_variants.tsv /data/Freyja_variants.tsv && cmp /data/test_depths.tsv /data/Freyja_depths.tsv" did not complete successfully: exit code: 1
```

I think this indicates that when you compare  /data/test_depths.tsv and /data/Freyja_depths.tsv, the files are no longer exactly the same. This is likely due to some new classifications between your current version and `1.3.4`.

jl32587 commented:
> This was the part of the test that failed:
> 
> ```
> #11 [test 3/4] RUN cmp /data/test_variants.tsv /data/Freyja_variants.tsv && cmp /data/test_depths.tsv /data/Freyja_depths.tsv
> #0 0.065 /data/test_variants.tsv /data/Freyja_variants.tsv differ: byte 139, line 1
> #11 ERROR: process "/usr/local/bin/_dockerfile_shell.sh cmp /data/test_variants.tsv /data/Freyja_variants.tsv && cmp /data/test_depths.tsv /data/Freyja_depths.tsv" did not complete successfully: exit code: 1
> ------
>  > [test 3/4] RUN cmp /data/test_variants.tsv /data/Freyja_variants.tsv && cmp /data/test_depths.tsv /data/Freyja_depths.tsv:
> #0 0.065 /data/test_variants.tsv /data/Freyja_variants.tsv differ: byte 139, line 1
> ------
> WARNING: local cache import at /tmp/.buildx-cache-freyja not found due to err: could not read /tmp/.buildx-cache-freyja/index.json: open /tmp/.buildx-cache-freyja/index.json: no such file or directory
> Dockerfile:62
> --------------------
>   60 |     
>   61 |     # Check validity of outputs
>   62 | >>> RUN cmp /data/test_variants.tsv /data/Freyja_variants.tsv && cmp /data/test_depths.tsv /data/Freyja_depths.tsv
>   63 |     RUN grep "Omicron" /data/test_demix.tsv && echo done
>   64 |     
> --------------------
> ERROR: failed to solve: process "/usr/local/bin/_dockerfile_shell.sh cmp /data/test_variants.tsv /data/Freyja_variants.tsv && cmp /data/test_depths.tsv /data/Freyja_depths.tsv" did not complete successfully: exit code: 1
> Error: buildx failed with: ERROR: failed to solve: process "/usr/local/bin/_dockerfile_shell.sh cmp /data/test_variants.tsv /data/Freyja_variants.tsv && cmp /data/test_depths.tsv /data/Freyja_depths.tsv" did not complete successfully: exit code: 1
> ```
> 
> I think this indicates that when you compare /data/test_depths.tsv and /data/Freyja_depths.tsv, the files are no longer exactly the same. This is likely due to some new classifications between your current version and `1.3.4`.

Any suggestions on how to fix this?

kapsakcj commented:
CC @kevinlibuit since he wrote the original dockerfile including the test layer

erinyoung commented:
Does updating lines 50-53 to your current version fix the problem?

If so, I recommend changing the hard-coded version to a variable.

jl32587 commented:
> Does updating lines 50-53 to your current version fix the problem?
> 
> If so, I recommend changing the hard-coded version to a variable.

So I need to create a new set of Freyja_variant.tsv and  Freyja_depth.tsv files?

erinyoung commented:
It looks like the two files used for comparison were download from Freyja's repo, so I don't think you need to create them. I'm hoping downloading the files that correspond with the version in question will resolve the issue.

jl32587 commented:
> It looks like the two files used for comparison were download from Freyja's repo, so I don't think you need to create them. I'm hoping downloading the files that correspond with the version in question will resolve the issue.

I have downloaded test data from the current version and edited line 50-53 & 60 to reflect the changes. I also replaced the '1.3.4' to the variable of current version. Hopefully it will work.

erinyoung commented:
Hmmm... your new files would need to be copied into the test layer.

Personally, I'm alright with just `cat <output files>` to show that the tests are working. I think @kapsakcj is right in suggesting feedback from @kevinlibuit .

jl32587 commented:
> Hmmm... your new files would need to be copied into the test layer.
> 
> Personally, I'm alright with just `cat <output files>` to show that the tests are working. I think @kapsakcj is right in suggesting feedback from @kevinlibuit .

Yeah, we'll see what he suggests. Thanks!

jl32587 commented:
> @jl32587 and @kevinlibuit , I adjusted the test so that it no longer compares the result files from an older version. Instead, it just outputs the contents of the files to stdout. Is this an okay change?
> 
> The multiple commits are because I couldn't get the filenames right.
> 
> I also fixed a label.

Thanks Erin! It looks great to me!

erinyoung commented:
I'm going to 
- merge this PR
- deploy to Dockerhub and quay with tag '1.3.12'
- overwrite the 'latest' tag
